lib/Semantics/expression.cpp

```diff
@@ -870,16 +872,25 @@ std::vector<Subscript> ExpressionAnalyzer::AnalyzeSectionSubscripts(
 }
 
 MaybeExpr ExpressionAnalyzer::Analyze(const parser::ArrayElement &ae) {
-  std::vector<Subscript> subscripts{AnalyzeSectionSubscripts(ae.subscripts)};
   if (MaybeExpr baseExpr{Analyze(ae.base)}) {
-    if (std::optional<DataRef> dataRef{ExtractDataRef(std::move(*baseExpr))}) {
-      if (!subscripts.empty()) {
-        return ApplySubscripts(std::move(*dataRef), std::move(subscripts));
+    if (ae.subscripts.empty()) {
+      // will be converted to function call later or error reported
+      return std::nullopt;
+    } else if (baseExpr->Rank() == 0) {
+      if (const Symbol * symbol{GetLastSymbol(*baseExpr)}) {
+        Say("'%s' is not an array"_err_en_US, symbol->name());
       }
+    } else if (std::optional<DataRef> dataRef{
+                   ExtractDataRef(std::move(*baseExpr))}) {
+      return ApplySubscripts(
+          std::move(*dataRef), AnalyzeSectionSubscripts(ae.subscripts));
     } else {
       Say("Subscripts may be applied only to an object, component, or array constant"_err_en_US);
     }
   }
+  // error was reported: analyze subscripts without reporting more errors
```

Comment:
I don't understand this comment.  Do you mean that an error may been reported?

Reply:
We only hit this line if an error was reported about the base of the array element. In that case we want to analyze the subscript expressions (because every expression should be analyzed) but not report any errors from that analysis because the might be confusing. E.g. in data01.f90, after "'persn' is not an array" you would get an error about 'Abcd Efgh' not being an integer (because array subscripts are required to be integers).